### PR TITLE
Focal Frequency Loss TF2 Keras implementation

### PR DESCRIPTION
**Pull request by sansyo**
Hello, thanks to publish your great idea on the web.

I tried to implement tensorflow 2(keras) version of Focal Frequency Loss.
But I noticed my loss value easily inflate into inf....... is there any way to fix that?

### CODE REVIEW

**Comment by Mark-Yu1** [comment]
When will you release the code?

**Comment by sansyo** [comment]
import tensorflow as tf

def fft2d(img):
img = tf.cast(img,dtype=tf.complex64)
fft = tf.signal.fft2d(img)
fft = tf.math.real(fft)
return fft

#loss stable than focal frequency loss
def frequency_loss(y_true,y_pred):

fft_true = fft2d(y_true)
fft_pred = fft2d(y_pred)

return tf.keras.losses.MeanSquaredError()(fft_true,fft_pred)


#easily go to inf....
def focal_frequency_loss(y_true,y_pred):

fft_true = fft2d(y_true)
fft_pred = fft2d(y_pred)

alpha = tf.constant(1.0, dtype=tf.float32)

sub = tf.math.subtract(fft_true,fft_pred)
sub_abs = tf.math.abs(sub)
norm = tf.math.divide(sub_abs, tf.math.reduce_max(sub_abs))

weight = tf.math.pow(norm,alpha)

se = tf.math.square(tf.math.subtract(fft_true,fft_pred))

w_se = tf.math.multiply(weight,se)


loss = tf.math.reduce_mean(w_se)


return loss

2021年9月2日(木) 16:20 Wangbo Yu ***@***.***>:
